repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -122,190 +136,145 @@ class ReposeRoutingServlet @Inject()(@Value(ReposeSpringProperties.CORE.REPOSE_V
+            fixLocationHeader(
+              servletRequest,
+              resp,
+              route,
+              target.url.getPath,
+              destination.getRootPath)
+          }
+        }
+      }.recover {
+        case e: IOException if ExceptionUtils.getRootCause(e).isInstanceOf[ReadLimitReachedException] =>
+          logger.error(RequestContentErrorMessage, e)
+          trySendError(resp, SC_REQUEST_ENTITY_TOO_LARGE, RequestContentErrorMessage)
+        case e: OriginServiceCommunicationException =>
+          logger.error(OriginServiceErrorMessage, e)
+          trySendError(resp, SC_SERVICE_UNAVAILABLE, OriginServiceErrorMessage)
         case e =>
           logger.error(FailedToRouteMessage, e)
           trySendError(resp, SC_INTERNAL_SERVER_ERROR, FailedToRouteMessage)
       }
+
       logger.trace("returning response...")
     }
   }
 
-  def trySendError(resp: HttpServletResponse, sc: Int, msg: String): Unit = {
-    if (!resp.isCommitted) {
-      resp.sendError(sc, msg)
+  def trySendError(servletResponse: HttpServletResponse, sc: Int, msg: String): Unit = {
+    if (!servletResponse.isCommitted) {
+      servletResponse.sendError(sc, msg)
     } else {
       logger.error(s"Failed to Send Error $sc with message '$msg' due to response already being committed")
     }
   }
 
-  def doRoute(servletRequest: HttpServletRequestWrapper, servletResponse: HttpServletResponse): Unit = {
-    val attributeDestinations = Option(servletRequest.getAttribute(CommonRequestAttributes.DESTINATIONS))
-      .map(_.asInstanceOf[List[RouteDestination]])
-      .getOrElse(List.empty)
-    val defaultDestinations = defaultDestination
-      .map(defDest => new RouteDestination(defDest.getId, servletRequest.getRequestURI, -1))
-      .toList
-    (attributeDestinations ++ defaultDestinations)
-      .sortWith(_.compareTo(_) > 0) // Highest Quality first
-      .headOption match {
-      case Some(routeDest) =>
-        localNode match {
-          case Some(node) =>
-            // todo: this can be deleted if we drop the WAR deployment (and thus, internal dispatching)
-            // todo: instead, a client from the http-client-service can be used directly
-            doDestination(servletRequest, servletResponse, routeDest, node)
-          case _ =>
-            logger.error("Invalid local node definition")
-            servletResponse.setStatus(SC_SERVICE_UNAVAILABLE)
-        }
-      case _ =>
-        logger.warn("No Valid routing destination specified.")
-        servletResponse.setStatus(SC_INTERNAL_SERVER_ERROR)
+  def getRoute(servletRequest: HttpServletRequest): Try[RouteDestination] = Try {
+    val attributeRoutes = Option(servletRequest.getAttribute(CommonRequestAttributes.DESTINATIONS)) match {
+      case None => Seq.empty
+      case Some(destinations: java.util.List[RouteDestination]) => destinations.asScala
+      case Some(destinations: Seq[RouteDestination]) => destinations
```

Comment:
What? Why would you have either a `List` or a `Seq`? Seems like we should know which it is.

Reply:
You're right, we do know. Prior to this story, the code was expecting a Scala `List`, which would not have worked since the destination selection filters currently put a `java.util.List` into the attributes. So here I just added support for both Java and Scala "lists" because it was easy to do and doesn't seem to harm anything.

If you would rather be strict about the type of list, that's easy enough to do.